New pool implementation: Cannot [request entity completed] when in state [WaitingForResponseDispatch]

E.g. here: https://jenkins.akka.io:8498/job/akka-http-nightly-release-10.0-against-akka-2.4/32/consoleFull

```
[ERROR] [01/15/2018 09:15:01.744] [HostConnectionPoolSpec-akka.actor.default-dispatcher-4] [akka.actor.ActorSystemImpl(HostConnectionPoolSpec)] [0 (WaitingForResponseDispatch)] Slot execution failed. That's probably a bug. Please file a bug at https://github.com/akka/akka-http/issues. Slot is restarted.
java.lang.IllegalStateException: Cannot [request entity completed] when in state [WaitingForResponseDispatch]
	at akka.http.impl.engine.client.pool.SlotState.illegalState(SlotState.scala:84)
	at akka.http.impl.engine.client.pool.SlotState.onRequestEntityCompleted(SlotState.scala:58)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot$$anonfun$onRequestEntityCompleted$1.apply(NewHostConnectionPool.scala:150)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot$$anonfun$onRequestEntityCompleted$1.apply(NewHostConnectionPool.scala:150)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.runOneTransition$1(NewHostConnectionPool.scala:175)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.loop$1(NewHostConnectionPool.scala:244)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.updateState(NewHostConnectionPool.scala:253)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.onRequestEntityCompleted(NewHostConnectionPool.scala:150)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection$$anonfun$5$$anonfun$apply$2.apply(NewHostConnectionPool.scala:344)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection$$anonfun$5$$anonfun$apply$2.apply(NewHostConnectionPool.scala:344)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection.withSlot(NewHostConnectionPool.scala:333)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection$$anonfun$5.apply(NewHostConnectionPool.scala:344)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection$$anonfun$5.apply(NewHostConnectionPool.scala:343)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$$anonfun$akka$http$impl$engine$client$pool$NewHostConnectionPool$HostConnectionPoolStage$$anon$$safely$1$$anonfun$apply$1.apply$mcV$sp(NewHostConnectionPool.scala:468)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$$anonfun$6.apply(NewHostConnectionPool.scala:467)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$$anonfun$6.apply(NewHostConnectionPool.scala:467)
```

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/NewHostConnectionPool.scala
@@ -155,6 +155,7 @@ private[client] object NewHostConnectionPool {
             updateState(_.onResponseReceived(this, response))
           def onResponseDispatchable(): Unit =
             updateState(_.onResponseDispatchable(this))
+
           def onResponseEntitySubscribed(): Unit =
             updateState(_.onResponseEntitySubscribed(this))
           def onResponseEntityCompleted(): Unit =
@@ -229,6 +230,7 @@ private[client] object NewHostConnectionPool {
                     cancelCurrentTimeout()
                     closeConnection()
                     state.onShutdown(this)
+                    logic.slotsWaitingForDispatch.remove(this)
                     OptionVal.None
                   } catch {
                     case NonFatal(ex) ⇒

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/SlotState.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/SlotState.scala
@@ -224,6 +224,7 @@ private[pool] object SlotState {
     override def onConnectionFailed(ctx: SlotContext, cause: Throwable): SlotState = this
     override def onConnectionAttemptFailed(ctx: SlotContext, cause: Throwable): SlotState = this
     override def onRequestEntityFailed(ctx: SlotContext, cause: Throwable): SlotState = this
+    override def onRequestEntityCompleted(ctx: SlotContext): SlotState = this
   }
 
   private[pool] /* to avoid warnings */ trait BusyWithResultAlreadyDispatched extends ConnectedState with BusyState {

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/SlotState.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/SlotState.scala
@@ -48,31 +48,31 @@ private[pool] sealed abstract class SlotState extends Product {
   def isIdle: Boolean
   def isConnected: Boolean
 
-  def onPreConnect(ctx: SlotContext): SlotState = illegalState(ctx, "preConnect")
-  def onConnectionAttemptSucceeded(ctx: SlotContext, outgoingConnection: Http.OutgoingConnection): SlotState = illegalState(ctx, "connected attempt succeeded")
-  def onConnectionAttemptFailed(ctx: SlotContext, cause: Throwable): SlotState = illegalState(ctx, "connection attempt failed")
+  def onPreConnect(ctx: SlotContext): SlotState = illegalState(ctx, "onPreConnect")
+  def onConnectionAttemptSucceeded(ctx: SlotContext, outgoingConnection: Http.OutgoingConnection): SlotState = illegalState(ctx, "onConnectionAttemptSucceeded")
+  def onConnectionAttemptFailed(ctx: SlotContext, cause: Throwable): SlotState = illegalState(ctx, "onConnectionAttemptFailed")
 
-  def onNewRequest(ctx: SlotContext, requestContext: RequestContext): SlotState = illegalState(ctx, "new request")
+  def onNewRequest(ctx: SlotContext, requestContext: RequestContext): SlotState = illegalState(ctx, "onNewRequest")
 
   /** Will be called either immediately if the request entity is strict or otherwise later */
-  def onRequestEntityCompleted(ctx: SlotContext): SlotState = illegalState(ctx, "request entity completed")
-  def onRequestEntityFailed(ctx: SlotContext, cause: Throwable): SlotState = illegalState(ctx, "request entity failed")
+  def onRequestEntityCompleted(ctx: SlotContext): SlotState = illegalState(ctx, "onRequestEntityCompleted")
+  def onRequestEntityFailed(ctx: SlotContext, cause: Throwable): SlotState = illegalState(ctx, "onRequestEntityFailed")
 
-  def onResponseReceived(ctx: SlotContext, response: HttpResponse): SlotState = illegalState(ctx, "receive response")
+  def onResponseReceived(ctx: SlotContext, response: HttpResponse): SlotState = illegalState(ctx, "onResponseReceived")
 
   /** Called when the response out port is ready to receive a further response (successful or failed) */
-  def onResponseDispatchable(ctx: SlotContext): SlotState = illegalState(ctx, "responseDispatched")
+  def onResponseDispatchable(ctx: SlotContext): SlotState = illegalState(ctx, "onResponseDispatchable")
 
-  def onResponseEntitySubscribed(ctx: SlotContext): SlotState = illegalState(ctx, "responseEntitySubscribed")
+  def onResponseEntitySubscribed(ctx: SlotContext): SlotState = illegalState(ctx, "onResponseEntitySubscribed")
 
   /** Will be called either immediately if the response entity is strict or otherwise later */
-  def onResponseEntityCompleted(ctx: SlotContext): SlotState = illegalState(ctx, "response entity completed")
-  def onResponseEntityFailed(ctx: SlotContext, cause: Throwable): SlotState = illegalState(ctx, "response entity failed")
+  def onResponseEntityCompleted(ctx: SlotContext): SlotState = illegalState(ctx, "onResponseEntityCompleted")
+  def onResponseEntityFailed(ctx: SlotContext, cause: Throwable): SlotState = illegalState(ctx, "onResponseEntityFailed")
 
-  def onConnectionCompleted(ctx: SlotContext): SlotState = illegalState(ctx, "connection completed")
-  def onConnectionFailed(ctx: SlotContext, cause: Throwable): SlotState = illegalState(ctx, "connection failed")
+  def onConnectionCompleted(ctx: SlotContext): SlotState = illegalState(ctx, "onConnectionCompleted")
+  def onConnectionFailed(ctx: SlotContext, cause: Throwable): SlotState = illegalState(ctx, "onConnectionFailed")
 
-  def onTimeout(ctx: SlotContext): SlotState = illegalState(ctx, "timeout")
+  def onTimeout(ctx: SlotContext): SlotState = illegalState(ctx, "onTimeout")
 
   def onShutdown(ctx: SlotContext): Unit = ()
 

